Migrating to baron scrollbar WIP

alexanderzobnin opened the pull request:
Trying to solve #11053 by migrating to  [baron](https://github.com/Diokuz/baron) scrollbar which doesn't affect scroll behaviour and only changes its styles.

Work in progress, need more tests on different browsers/OSs/platforms.

alexanderzobnin committed:
Initially move to baron scrollbar

alexanderzobnin committed:
scrollbar: fix legend rendering issues

alexanderzobnin committed:
scrollbar: fix Firefox issue (white stripe on the right of scrollbar)

daniellee committed:
Merge branch 'fix-11053' of https://github.com/alexanderzobnin/grafana into alexanderzobnin-fix-11053

alexanderzobnin committed:
scrollbar: fix side menu on mobile devices

alexanderzobnin committed:
scrollbar: fix Firefox scroll position restore

daniellee committed:
Merge branch 'fix-11053' of https://github.com/alexanderzobnin/grafana into alexanderzobnin-fix-11053

daniellee commented:
The menu seems to mostly work and scroll position seems fine on Firefox and iOS.

I found a couple of things to fix:

- Baron is not added to yarn.lock. Needs to be added with `yarn add baron`.

- In Chrome and Firefox, the width of the dashboard seems to be incorrect - too wide. Getting horizontal scroll - scrolled to the left:

    ![image](https://user-images.githubusercontent.com/434655/38309912-79afc8b2-381b-11e8-96ed-85791f0cb2bd.png)

    scrolled to the right:

    ![image](https://user-images.githubusercontent.com/434655/38310266-6ad238c4-381c-11e8-94c9-9ec33baaf71b.png)

- Getting the error: `legendScrollbar.destroy is not a function` on small screens when a clicking a link in the mobile menu. When testing on iOS - sometimes getting an empty black screen when clicking on a link in the menu.

alexanderzobnin commented:
Thanks for review, I also noticed these issues so working on fixes.

alexanderzobnin committed:
scrollbar: fix 'legendScrollbar.destroy is not a function' error

alexanderzobnin committed:
scrollbar: fix dashboard width bug

alexanderzobnin committed:
scrollbar: remove perfect-scrollbar and add baron to package list

daniellee committed:
Merge branch 'fix-11053' of https://github.com/alexanderzobnin/grafana into alexanderzobnin-fix-11053

alexanderzobnin committed:
scrollbar: fix dashboard width updating for different modes

daniellee commented:
@alexanderzobnin found two issues. There is no scroll on the search in the mobile view:

![image](https://user-images.githubusercontent.com/434655/38322256-3611c3d4-383a-11e8-81f1-086754146d0b.png)

The other issue is that on iOS, sometimes after choosing a page from the menu, it does not load. Trying to find out why. It does not happen on the master branch.

daniellee commented:
![image](https://user-images.githubusercontent.com/434655/38323437-88c15100-383d-11e8-8257-50d7655d4263.png)

Can reproduce this in Chrome sometimes, though not as often as on an iPhone. On the grafana-scrollbar div, sometimes the max-width and width get set to 0 which hides all the content.

marefr commented:
First general concern/question. Is it only my that have problems in Chrome that I can't scroll sometimes, but clicking on some area (giving element focus I suppose) allow me to scroll again. This is really irritating. Maybe my laptop's scroll having issues? 

Second general concern: Seems like my homepage dashboard no longer can render starred/recent dashboards and installed apps panels. Possibly no height?

Confirmed that this makes #11185 almost completely solved, at least when using an extension called FireShot in Chrome which basically scrolling the page down. Think this is how most of these extensions worls. Partially because there's still an issue when the sidemenu are visible (see first screenshot), but works very good when in kiosk mode (see second screenshot)

![fireshot capture 22 - grafana - big dashboard_ - http___localhost_3000_d_000000003_big-dashboard](https://user-images.githubusercontent.com/1668778/38339945-28c06550-3871-11e8-91b2-3c44bad74b26.png)

![fireshot capture 16 - grafana - big dashboard_ - http___localhost_3000_d_000000003_big-dashboard](https://user-images.githubusercontent.com/1668778/38339946-28e1aa1c-3871-11e8-8524-84caf7730266.png)

torkelo commented on public/app/core/components/scroll/scroll.ts:
```diff
@@ -31,13 +60,25 @@ export function geminiScrollbar() {
       );
 
+      // force updating dashboard width
+      appEvents.on('toggle-sidemenu', forceUpdate);
```
this could be a memory leak, no unsubscribe  (or scope passed as last argument to unsubscribe on scope destory)

alexanderzobnin replied:
Ah, yea, missed that.

torkelo commented:
Good work alex! 

Tested this and not sure if this is a remaining issue or not but the legend height is buggy (to high). 

<img width="813" alt="screen shot 2018-04-05 at 09 07 04" src="https://user-images.githubusercontent.com/10999/38351651-bd9e8c26-38b0-11e8-86e8-16fcb7d88457.png">

alexanderzobnin committed:
scrollbar: fix potential memory leaks in event handlers

alexanderzobnin committed:
scrollbar: fix graph legend height

alexanderzobnin committed:
Merge remote-tracking branch 'upstream/master' into fix-11053

alexanderzobnin committed:
scrollbar: fix search scroller in mobile view

daniellee committed:
Merge branch 'fix-11053' of https://github.com/alexanderzobnin/grafana into alexanderzobnin-fix-11053

alexanderzobnin committed:
scrollbar: styles cleanup

alexanderzobnin commented:
I still have some issues:

- [ ]  White stripes on the right of scrolling element in Firefox
- [ ]  When you scrolling graph legend by mouse wheel and reach the top or bottom, dashboard isn't scrolling until you stop and wait for a moment (I think it's due to hardcoded 300 ms for event debounce https://github.com/Diokuz/baron/blob/master/src/autoUpdate.js#L40)

alexanderzobnin committed:
scrollbar: fix add panel height bug

torkelo commented:
Looks good.

1) PNG (phantomjs) rendering is not working at all for no panel type
2) not a show stopper for PR, but page down/up does not work after page load for me, I have to click somewhere in the page (probably to give scroll conainer focus) for page up/down to work.

alexanderzobnin committed:
scrollbar: fix phantomjs rendering error

daniellee committed:
Merge branch 'fix-11053' of https://github.com/alexanderzobnin/grafana into alexanderzobnin-fix-11053

daniellee committed:
scrollbar: use native scroll for page

daniellee committed:
Merge remote-tracking branch 'upstream/master' into 11053-remove-native-scrollbar

marefr commented:
Still things left to fix:
1. PNG (phantomjs) rendering is not working for graph panel having legends to the right
2. Starred/recent dashboards and installed apps panel doesn't render correctly

![image](https://user-images.githubusercontent.com/1668778/38616829-9e4b1116-3d95-11e8-989b-359496d99a92.png)

marefr commented:
We have a new branch now as well called 11053-remove-native-scrollbar where we've added the native scrollbar to all pages.

marefr committed:
panel: add baron scroller to correct element

This resolves issue with alert list panel getting scrollbars
attached to incorrect element. Now the panel content are
rendered correctly and all content are displayed as
expected.

marefr committed:
fix so that dash list panel are rendered correctly

This resolves issue with dash list panel getting scrollbars
    attached to incorrect elements. Now the panel content are
    rendered correctly and all content are displayed as
    expected.

marefr committed:
fix so that page scrollbars can be scrolled by keyboard on page load

The page scrollbars are custom, not rendered on the body element and
with css property overflow set for scroll to be enabled.
For being able to scroll the page using the keyboard when a
page loads, some custom code was needed.
This fix should both work when doing a full reload of a url and when
navigating to other pages/dashboards.
For those pages having an input field that are focused on load,
scrolling by keyboard (arrow up/down) will obviously not work.

marefr committed:
dashboard: show baron scrollbar in dashboard panel when mouse is over

This should make the scrolling user experience much better since it
will highlight and show that content actually can be scrolled when
hovering over a panel.

daniellee committed:
Merge remote-tracking branch 'upstream/11053-remove-native-scrollbar' into 11053-remove-native-scrollbar

daniellee committed:
scrollbar: remove unused div

Fixes the height of the settings menu

marefr committed:
minor scrollbar fixes

Hide page scrollbar when rendering using phantomjs.
Hide baron scrollbar when rendering using phantomjs.

daniellee committed:
scrollbar: fixes continuation scrolling for iOS

The -webkit-overflow-scrolling is an iOS only property:

https://developer.mozilla.org/en-US/docs/Web/CSS/-webkit-overflow-scrolling

that turns on momementum/continuation scrolling for iOS devices. This means
that when swiping, the scroll continues for a half second rather than instantly
stopping the scroll when the user lifts their finger from the screen.

daniellee committed:
scrollbar: fix so no overflow for legend under graph

Adds 1px of padding to the graph legend scroll div so that a non-table
legend does not get an unnecessary scroll bar.

alexanderzobnin committed:
fix right side legend rendering in phantomjs

alexanderzobnin committed:
Merge remote-tracking branch 'upstream/11053-remove-native-scrollbar' into 11053-remove-native-scrollbar

alexanderzobnin committed:
remove mistakenly added styles

daniellee reviewed (approved):
Fantastic work by Alex to get this working. Loads of tricky edge cases (phantomjs, firefox etc.) that needed to be solved.
